pyomo/contrib/doe/example/reactor_kinetics.py

```diff
@@ -29,209 +29,199 @@
-    m.kp2 = pyo.Var(m.scena, m.t, initialize=kp2_init)
-
-
-    def T_control(m,t):
-        """
-        T at interval timepoint equal to the T of the control time point at the beginning of this interval
-        Count how many control points are before the current t;
-        locate the nearest neighbouring control point before this t
-        """
-        if t in m.t_con:
-            return pyo.Constraint.Skip
-        else:
-            j = -1 
-            for t_con in m.t_con:
-                if t>t_con:
-                    j+=1
-            neighbour_t = t_control[j]
-            return m.T[t] == m.T[neighbour_t]
-        
-    
-    def cal_kp1(m,z,t):
-        """
-        Create the perturbation parameter sets 
-        m: model
-        z: scenario number
-        t: time
-        """
-        # LHS: 1/h
-        # RHS: 1/h*(kJ/mol *1000J/kJ / (J/mol/K) / K)
-        return m.kp1[z,t] == m.A1[z]*pyo.exp(-m.E1[z]*1000/(m.R*m.T[t])) 
+        para_list = ['A1', 'A2', 'E1', 'E2']
+
+        if not control_time:
+            control_time = [0,0.125,0.25,0.375,0.5,0.625,0.75,0.875,1]
             
-    def cal_kp2(m,z,t):
-        """
-        Create the perturbation parameter sets 
-        m: model
-        z: m.pert, upper or normal or lower perturbation
-        t: time
-        """
-        # LHS: 1/h
-        # RHS: 1/h*(kJ/mol *1000J/kJ / (J/mol/K) / K)
-        return m.kp2[z,t] == m.A2[z]*pyo.exp(-m.E2[z]*1000/(m.R*m.T[t])) 
+        if not control_val:
+            control_val = [300]*9
+
+        controls = {}
+        for i, t in enumerate(control_time):
+            controls[t]=control_val[i]
         
-    def dCdt_control(m,z,y,t):
-        """
-        Calculate CA in Jacobian matrix analytically 
-        z: scenario No.
-        y: CA, CB, CC
-        t: timepoints
-        """
-        if y=='CA':
-            return m.dCdt[z,y,t] == -m.kp1[z,t]*m.C[z,'CA',t]    
-        elif y=='CB':
-            return m.dCdt[z,y,t] == m.kp1[z,t]*m.C[z,'CA',t] - m.kp2[z,t]*m.C[z,'CB',t]
-        elif y=='CC':
-            return m.dCdt[z,y,t] == m.kp2[z,t]*m.C[z,'CB',t]
+        ### Add variables 
+        mod.CA_init = CA_init
+        mod.para_list = para_list
+        t_control = control_time
         
-    def alge(m,z,t):
-        """
-        The algebraic equation for mole balance
-        z: m.pert
-        t: time
-        """
-        return m.C[z,'CA',t] + m.C[z,'CB',t] + m.C[z,'CC', t] == m.CA0[0]
-    
+        # timepoints
+        mod.t = ContinuousSet(bounds=(t_range[0], t_range[1]))
         
-    # Control time
-    m.T_rule = pyo.Constraint(m.t, rule=T_control)
-    
-    # calculating C, Jacobian, FIM
-    m.k1_pert_rule = pyo.Constraint(m.scena, m.t, rule=cal_kp1)
-    m.k2_pert_rule = pyo.Constraint(m.scena, m.t, rule=cal_kp2)
-    m.dCdt_rule = pyo.Constraint(m.scena,m.C_set, m.t, rule=dCdt_control)
+        # Control time points
+        mod.t_con = pyo.Set(initialize=t_control)
```

Comment:
This seems redundant with the earlier code defining mod.t_con. Can this be streamlined?

Reply:
streamlined by moving some common codes to the global part